From 5b134f1c98c85ca452ee28d13e5423e4ccd91d91 Mon Sep 17 00:00:00 2001
From: Igal Liberman <igall@marvell.com>
Date: Wed, 22 Mar 2017 16:20:56 +0200
Subject: [PATCH 0912/1345] crypto: inside-secure: optimize DSE configuration

commit  cc99ed1498054eb16a1dd59a3a8d06d0f4b1c58c from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Enable enable_single_wr option in the DSE (Data Storage Engine).

When enable_single_wr is not enabled, the DSE will only write those
parts of a result descriptor that need updating, which means a final
result descriptor will be written in 2 or 3 smaller transfers.
When enable_single_wr is enabled the DSE will combine these 2-3
updates into one large write transfer,
generally improving performance.

Change-Id: Ibdd4fd139eef776f658df13ecba5d86a57639d98
Signed-off-by: Igal Liberman <igall@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37851
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/crypto/inside-secure/safexcel.c |    1 +
 drivers/crypto/inside-secure/safexcel.h |    1 +
 2 files changed, 2 insertions(+)

diff --git a/drivers/crypto/inside-secure/safexcel.c b/drivers/crypto/inside-secure/safexcel.c
index 121f12d..7b045cd 100644
--- a/drivers/crypto/inside-secure/safexcel.c
+++ b/drivers/crypto/inside-secure/safexcel.c
@@ -534,6 +534,7 @@ static int eip197_hw_init(struct device *dev, struct safexcel_crypto_priv *priv)
 	val |= EIP197_HIA_DxE_CFG_MIN_DATA_SIZE(7) | EIP197_HIA_DxE_CFG_MAX_DATA_SIZE(8);
 	val |= EIP197_HIA_DxE_CFG_DATA_CACHE_CTRL(WR_CACHE_3BITS);
 	val |= EIP197_HIA_DSE_CFG_BUFFER_CTRL(2);
+	val |= EIP197_HIA_DSE_CFG_EN_SINGLE_WR;
 	writel(val, EIP197_HIA_DSE(priv) + EIP197_HIA_DSE_CFG);
 
 	/* Leave the DSE threads reset state */
diff --git a/drivers/crypto/inside-secure/safexcel.h b/drivers/crypto/inside-secure/safexcel.h
index b13ec84..9248f3d 100644
--- a/drivers/crypto/inside-secure/safexcel.h
+++ b/drivers/crypto/inside-secure/safexcel.h
@@ -302,6 +302,7 @@
 #define EIP197_HIA_DFE_CFG_DIS_DEBUG		(BIT(31) | BIT(29))
 #define EIP197_HIA_DSE_CFG_DIS_DEBUG		BIT(31)
 #define EIP197_HIA_DSE_CFG_BUFFER_CTRL(n)	(((n) & 0x3) << 14)
+#define EIP197_HIA_DSE_CFG_EN_SINGLE_WR		BIT(29)
 
 /* EIP197_HIA_DFE/DSE_THR_CTRL */
 #define EIP197_DxE_THR_CTRL_EN			BIT(30)
-- 
1.7.9.5

